### PR TITLE
feat(sdk): support compressed response

### PR DESCRIPTION
**Pull request by mask-pp**
## 1. Purpose or design rationale of this PR

Enable use compression algorithm, support: gzip, zlib(deflate)
i.g:

- use curl:

`
curl --location 'http://localhost:8545' -H 'Content-Type: application/json' --compressed --data '{"jsonrpc": "2.0", "method": "scroll_getBlockTraceByNumberOrHash","params": ["0x2ca4"],"id": 1}' -v -w '%{size_download}'
`

- use by SDK:

`
client, err := ethclient.Dial("http://localhost:8545")

assert.NoError(t, err)

client.SetHeader("Accept-Encoding", "gzip")
`

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [x] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [x] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

**Comment by Thegaram** [comment]
Are you sure this is not enabled by default? I've used queries like this before (using `--compressed`):

```
curl -H 'Content-Type: application/json' --compressed -d '{"jsonrpc":"2.0","id":"id","method":"eth_getLogs", "params": [{ "fromBlock": "0x0", "topics": [] }]}' http://localhost:8545
```

**Comment by mask-pp** [comment]
> Are you sure this is not enabled by default? I've used queries like this before (using `--compressed`):
> 
> ```
> curl -H 'Content-Type: application/json' --compressed -d '{"jsonrpc":"2.0","id":"id","method":"eth_getLogs", "params": [{ "fromBlock": "0x0", "topics": [] }]}' http://localhost:8545
> ```

Use `--header 'Content-Encoding: gzip'` can works.

**Comment by mask-pp** [comment]
> Are you sure this is not enabled by default? I've used queries like this before (using `--compressed`):
> 
> ```
> curl -H 'Content-Type: application/json' --compressed -d '{"jsonrpc":"2.0","id":"id","method":"eth_getLogs", "params": [{ "fromBlock": "0x0", "topics": [] }]}' http://localhost:8545
> ```

So, we just are able to decode the compressed content if client set "Accept-Encoding: gzip".

File: rpc/http.go [hunk 1]
```diff
@@ -198,7 +201,9 @@ func (hc *httpConn) doRequest(ctx context.Context, msg interface{}) (io.ReadClos
 			Body:       body,
 		}
 	}
-	return resp.Body, nil
+
+	// if encoding is set use it.
+	return newDecodeCompression(req.Header.Get("Accept-Encoding"), resp.Body)
```
Comment (Thegaram) [inline rpc/http.go]:
```suggestion
	return newDecodeCompression(resp.Header.Get("Content-Encoding"), resp.Body)
```

File: rpc/http.go [hunk 2]
```diff
@@ -208,6 +213,28 @@ type httpServerConn struct {
...
+	tps := strings.Split(strings.TrimSpace(strings.ToLower(decoding)), ",")
+	var res io.ReadCloser
+	switch tps[0] {
+	case "gzip":
+		gz, err := gzip.NewReader(rc)
+		if err != nil {
+			return nil, err
+		}
+		res = gz
+	case "zlib", "deflate":
```
Comment (Thegaram) [inline rpc/http.go]: Based on [this](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding), `zlib` does not seem to be a standard encoding. (`"deflate"` means zlib, not `"zlib"`)

Comment (mask-pp) [reply rpc/http.go]: fixed

**Review by Thegaram — APPROVED** [review]

**Review by icemelon — APPROVED** [review]

**Pull request merged**